nexus/src/sagas.rs

```diff
@@ -184,18 +184,116 @@ async fn sic_alloc_server(
         .map_err(ActionError::action_failed)
 }
 
-async fn sic_create_network_interface(
+async fn sic_create_network_interfaces(
     sagactx: ActionContext<SagaInstanceCreate>,
-) -> Result<NetworkInterface, ActionError> {
+) -> Result<Option<Vec<NetworkInterface>>, ActionError> {
+    match sagactx.saga_params().create_params.network_interface {
+        params::InstanceNetworkInterfaceAttachment::None => Ok(None),
+        params::InstanceNetworkInterfaceAttachment::Default => {
+            sic_create_default_network_interface(&sagactx).await
+        }
+        params::InstanceNetworkInterfaceAttachment::Attach(
+            ref attach_params,
+        ) => sic_attach_network_interface(&sagactx, attach_params).await,
+        params::InstanceNetworkInterfaceAttachment::Create(
+            ref create_params,
+        ) => {
+            sic_create_custom_network_interfaces(&sagactx, create_params).await
+        }
+    }
+}
+
+/// Attach one or more existing network interfaces to the provided instance.
+async fn sic_attach_network_interface(
+    _sagactx: &ActionContext<SagaInstanceCreate>,
+    _attach: &params::InstanceAttachNetworkInterface,
+) -> Result<Option<Vec<NetworkInterface>>, ActionError> {
+    todo!();
+}
+
+/// Create one or more custom (non-default) network interfaces for the provided
+/// instance.
+async fn sic_create_custom_network_interfaces(
+    sagactx: &ActionContext<SagaInstanceCreate>,
+    interface_params: &params::InstanceCreateNetworkInterface,
+) -> Result<Option<Vec<NetworkInterface>>, ActionError> {
     let osagactx = sagactx.user_data();
-    let params = sagactx.saga_params();
+    let saga_params = sagactx.saga_params();
     let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+    let vpc = osagactx
+        .datastore()
+        .vpc_fetch_by_name(
+            &saga_params.project_id,
+            &db::model::Name::from(interface_params.vpc_name.clone()),
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
 
+    let mut interfaces =
+        Vec::with_capacity(interface_params.interface_params.len());
+    for params in interface_params.interface_params.iter() {
+        // TODO-correctness: It seems racy to fetch the subnet and create the
```

Comment:
Related to this - when VPC subnets get deleted, do we ensure that there aren't any NetworkInterfaces using them?

Reply:
No, we don't currently do that check when deleting any given subnet. I don't _think_ that's a huge problem here. The saga will still unwind any resources it creates, the interface(s) in this case.

But in general, this is a bit issue. @davepacheco has added a number of comments in the Nexus sources for it. There are quite a few cases where we currently do multiple database requests. For reasons of correctness or performance, we should implement those with more complex queries (joins or subqueries). I think _simple_ JOINs shouldn't be too bad with Diesel (I've been writing a few to get the sleds associated with interfaces, for example). I don't know about subqueries, and I don't think it's possible to write CTEs at all with the normal Diesel API (one has to implement the `QueryFragment` trait we have here instead).

I don't see a TODO in the `Datastore::vpc_delete_subnet` method for checking for interfaces before deleting now, but I think there are a _lot_ of similar cases we'll need to consider. For example, we don't detach a NIC from an instance when that instance is deleted; or remove routes for a subnet when deleting the subnet; or update; or check for instances (or any child resources) when deleting a project. Some of these have TODOs associated with them, some don't. I'm not sure a tracking issue makes sense now, but I definitely don't want to lose them.

All that's to say -- I agree it's a big issue; I think we should do a thorough review of the DB layer at some point; and I'm not sure how to track all this stuff :)

Reply:
So, I've been thinking a bit more about this. Most queries that delete a resource look like:

```rust
diesel::update(table_name)
    .filter(dsl::id.eq(resource_id)
    .set(dsl::time_deleted.eq(Utc::now());
```

That's nice and compact. But when there's a related resource that prevents deletion, such as an attached instance that should prevent deletion of a network interface, this won't quite work.

I mean, it would mechanically work to write that query. Diesel will execute it. We can even have a check in there that only updates records where the `instance_id.eq(None)`, for example, to prevent deleting a NIC that has a non-NULL instance. But we then can't tell the difference between a record with that primary key not existing, and not deleting one because the filtering clause removed the one that still had an instance associated with it. That is, if we learn that we deleted zero rows, we don't know if that's because there's no such NIC, or because a NIC was there, but was attached to an instance.

This is the reason for a lot of the skullduggery in the queries in this commit, and the ones in `collection_insert.rs`. We need to distinguish one of several kinds of errors, so we generate a query that should fail in some particular ways that disambiguates them. I'm a bit worried that, because we've got a lot of such constraints (lots of related resources), many of the queries will need to be hand-written in this fashion. It's a lot of work, and kinda defeats the purpose of Diesel in the first place.

There's nothing really actionable here. I think I just wanted to get some of these thoughts out, and make clear that we may need to write specialized queries to handle the various edge cases we need to enforce our application invariants. Transactions are an alternative, but one I think we'd rather avoid unless writing/testing/maintaining all these big queries is too much work. There may be others.

Reply:
> So, I've been thinking a bit more about this. Most queries that delete a resource look like:
> 
> ```rust
> diesel::update(table_name)
>     .filter(dsl::id.eq(resource_id)
>     .set(dsl::time_deleted.eq(Utc::now());
> ```
> 
> That's nice and compact. But when there's a related resource that prevents deletion, such as an attached instance that should prevent deletion of a network interface, this won't quite work.
> 
> I mean, it would mechanically work to write that query. Diesel will execute it. We can even have a check in there that only updates records where the `instance_id.eq(None)`, for example, to prevent deleting a NIC that has a non-NULL instance. But we then can't tell the difference between a record with that primary key not existing, and not deleting one because the filtering clause removed the one that still had an instance associated with it. That is, if we learn that we deleted zero rows, we don't know if that's because there's no such NIC, or because a NIC was there, but was attached to an instance.

Isn't this exactly what [update_and_check](https://github.com/oxidecomputer/omicron/blob/main/nexus/src/db/update_and_check.rs) solves? It lets us still access the underlying resource even if the query fails, and return a custom error message - we use this during [disk deletion](https://github.com/oxidecomputer/omicron/blob/755b07d3d13227b4ea192cea0650f7af47c8a15f/nexus/src/db/datastore.rs#L1610-L1637) to issue a query with a bunch of `.filters`, but provide a more refined error message on failure.

Reply:
That's awesome, thanks for pointing that out! I think that's a great place to start. I can't tell if it captures the semantics we need to prevent deletion of subnets if an interface in that subnet exists, or deleting a NIC that's still associated with an instance. The `update_and_check` stuff looks like it fails a query if we attempt to update a primary key that doesn't exist. The two cases above are different. In the first case, we'd need to fail an update to the `vpc_subnet` table, if a row referencing that in the `network_interface` table exists. In the second, fail the query if the `instance_id` column is non-NULL.

It might be possible to express those in the existing code, or we might be able to extend it easily to check more general conditions, possibly across tables. It's also possible I don't understand the whole thing :)

Reply:
Ah, no, I think I'm still wrong. We get back the found row, even in the case where the update failed. So we can look at the `time_deleted` and `instance_id` column to disambiguate (1) and (2) here. Thanks @smklein for pointing me to the `disk_delete` example, that was very helpful.

Reply:
Tracking this in #738